pkg/ostree/client.js

```diff
@@ -0,0 +1,513 @@
+                if (transaction_client) {
+                    if (subscription)
+                        subscription.remove();
+
+                    $(transaction_client).off();
+                    transaction_client.close();
+                }
+                transaction_client = null;
+                subscription = null;
+                trigger_changed();
+            }
+
+            function fail(ex) {
+                dp.reject(ex);
+                cleanup();
+            }
+
+            self.connect()
+                .fail(fail)
+                .done(function () {
+                    var proxy = self.get_os_proxy(os);
+
+                    if (!proxy)
+                        return fail(cockpit.format(_("OS $0 not found"), os));
+
+                    proxy.call(method, method_args)
+                        .fail(fail)
+                        .done(function(result) {
+                            var connect_args = {
+                                "superuser" : true,
+                                "address": result[0],
+                                "bus": "none"
+                            };
+
+                            transaction_client = cockpit.dbus(null, connect_args);
+                            $(transaction_client).on("close", function(event, ex) {
+                                fail(ex);
+                            });
+
+                            subscription = transaction_client.subscribe({ 'path' : "/", },
+                                function(path, iface, signal, args) {
+                                    if (signal == "DownloadProgress") {
+                                        var line = build_progress_line(args);
+                                        if (line)
+                                            dp.notify(line);
+                                    } else if (signal == "Message") {
+                                        dp.notify(args[0]);
+                                    } else if (signal == "Finished") {
+                                        if (args) {
+                                            if (args[0]) {
+                                                dp.resolve(args[1]);
+                                                cleanup();
+                                            } else {
+                                                fail(args[1]);
+                                            }
+                                        } else {
+                                            console.warn("Unexpected transaction response", args);
+                                            fail({ "problem": "protocol-error" });
+                                        }
+                                    }
+                                });
+                            transaction_client.call("/", TRANSACTION, "Start");
+                        });
+                });
+
+            return dp;
+        };
+    }
+
+    return new RPMOSTreeDbusClient();
```

Comment:
We haven't done this before. But now come to think of it, it's a great way to do a singleton without the nonsense we do in kubernetes/client.js

Are there any downsides that you know about?